src/libopensc/pkcs15-openpgp.c

```diff
@@ -296,14 +292,20 @@ sc_pkcs15emu_openpgp_init(sc_pkcs15_card_t *p15card)
 			prkey_info.usage          = key_cfg[i].prkey_usage;
 			prkey_info.native         = 1;
 			prkey_info.key_reference  = i;
-			prkey_info.modulus_length = bebytes2ushort(cxdata + 1);
 
 			strlcpy(prkey_obj.label, key_cfg[i].label, sizeof(prkey_obj.label));
 			prkey_obj.flags = SC_PKCS15_CO_FLAG_PRIVATE | SC_PKCS15_CO_FLAG_MODIFIABLE;
 			prkey_obj.auth_id.len      = 1;
 			prkey_obj.auth_id.value[0] = key_cfg[i].prkey_pin;
 
-			r = sc_pkcs15emu_add_rsa_prkey(p15card, &prkey_obj, &prkey_info);
+			if (cxdata[0] == 0x01) { /* RSA */
+				prkey_info.modulus_length = bebytes2ushort(cxdata + 1);
```

Comment:
Just to make sure: You mean something like
```
if (cxdata[0] == SC_OPENPGP_KEYALGO_RSA && r > 1) {
```
right?

Reply:
Almost.

You removed the length check after the call to read_file() a few lines earlier, and only keep the check whether the read failed. But the length read still may be too small (e.g. 0).

Looking at the whole section, I think there are more issues
1) the check for `cxdata[0] == 0` in line 289.
    If the length returned by read_file() is 0, then cxdata[]'s contents are undefined.
    (AFAIK only `static` variables get auto-initialized per the C standard)
    So for that check to be meaningful even in that case, I'd add a 
   ```
   memset(cxdata, 0, sizeof(cxdata));
   ```
   at around line 275.
2) the length check in line 301
    I'd use 
   ```
   if (cxdata[0] == SC_OPENPGP_KEYALGO_RSA && r >= 3) {
   ```
   because cxdata[]'s contents, that are used by the `bebytes2ushort()` call, are valid only if  3 bytes or more have been read.

The same applies to the logic for the public keys a few lines later.

Maybe a silly question:
The loops for private and public keys look very similar; is there a reason they are kept as two separate loops?

Reply:
Mh, it seems that my comment is somehow lost. Anyway. I think I fixed the issues.

> Maybe a silly question:
The loops for private and public keys look very similar; is there a reason they are kept as two separate loops?

Actually, I needed to think about this as well. Then I saw the comment above the first loop:

```
/* XXX: check if "halfkeys" can be stored with gpg2. If not, add keypairs in one loop */
```

I'd like to keep it like it is for now.